iree/compiler/Dialect/Flow/Transforms/test/identify_dispatch_regions2_hlo.mlir

```diff
@@ -16,65 +16,40 @@ func @simpleMath(%arg0 : tensor<4xf32>) -> tensor<4xf32> {
 
 // -----
 
-// CHECK-LABEL: @hloElementwiseOps
-func @hloElementwiseOps(%arg0 : tensor<4xf32>) -> tensor<4xf32> {
-  // CHECK-NEXT: %[[WORKLOAD:.+]] = constant 4
-  // CHECK-NEXT: %0 = flow.dispatch.region
-  // CHECK-SAME: [%[[WORKLOAD]] : index]
-  // CHECK-SAME: (%arg1 = %arg0 : tensor<4xf32>) -> tensor<4xf32> {
-  // CHECK-NEXT:   %1 = xla_hlo.add %arg1, %arg1 : tensor<4xf32>
-  %0 = xla_hlo.add %arg0, %arg0 : tensor<4xf32>
-  // CHECK-NEXT:   %2 = xla_hlo.subtract %1, %arg1 : tensor<4xf32>
-  %1 = xla_hlo.subtract %0, %arg0 : tensor<4xf32>
-  // CHECK-NEXT:   %3 = xla_hlo.multiply %2, %arg1 : tensor<4xf32>
-  %2 = xla_hlo.multiply %1, %arg0 : tensor<4xf32>
-  // CHECK-NEXT:   flow.return %3 : tensor<4xf32>
-  // CHECK-NEXT: }
-  // CHECK-NEXT: return %0 : tensor<4xf32>
-  return %2 : tensor<4xf32>
-}
-
-// -----
-
-// CHECK-LABEL: @interleavedDot
-func @interleavedDot(%arg0 : tensor<4x4xf32>) -> tensor<4x4xf32> {
+// CHECK-LABEL: @isolatedDot
+func @isolatedDot(%arg0 : tensor<4x4xf32>) -> tensor<4x4xf32> {
   // NOTE: Fragile ordering. Workload constants are emitted in order a the
   // top of the block.
-  // CHECK: %[[WORKLOAD0:.+]] = constant 16 : index
-  // CHECK: %[[R0:.+]] = flow.dispatch.region
-  // CHECK-SAME: [%[[WORKLOAD0]] : index]
-  // CHECK-SAME: (%arg1 = %arg0 : tensor<4x4xf32>) -> tensor<4x4xf32> {
-  // CHECK-NEXT:   %[[DADD:.*]] = xla_hlo.add %arg1, %arg1 : tensor<4x4xf32>
-  // CHECK-NEXT:   %[[DDOT:.*]] = "xla_hlo.dot"(%[[DADD]], %arg1)
-  // CHECK-NEXT:   %[[DMUL:.*]] = xla_hlo.multiply %[[DDOT]], %arg1
-  // CHECK-NEXT: flow.return %[[DMUL]] : tensor<4x4xf32>
-  // CHECK-NEXT: }
+  // CHECK: flow.dispatch.region
+  // CHECK:   xla_hlo.add
+  // CHECK: flow.dispatch.region
+  // CHECK:   "xla_hlo.dot"
+  // CHECK: flow.dispatch.region
+  // CHECK:   xla_hlo.multiply
   %0 = xla_hlo.add %arg0, %arg0 : tensor<4x4xf32>
   %1 = "xla_hlo.dot"(%0, %arg0) : (tensor<4x4xf32>, tensor<4x4xf32>) -> tensor<4x4xf32>
   %2 = xla_hlo.multiply %1, %arg0 : tensor<4x4xf32>
-  // CHECK-NEXT: return %[[R0]] : tensor<4x4xf32>
   return %2 : tensor<4x4xf32>
 }
 
 // -----
 
-// CHECK-LABEL: func @caller
-func @caller(%arg0 : tensor<4xf32>) -> tensor<4xf32> {
-  // CHECK-NEXT: %[[WORKLOAD0:.+]] = constant 4 : index
-  // CHECK-NEXT: %[[R0:.+]] = flow.dispatch.region
-  // CHECK-SAME: [%[[WORKLOAD0]] : index]
-  // CHECK-SAME: (%arg1 = %arg0 : tensor<4xf32>) -> tensor<4xf32> {
-  // CHECK-NEXT:   %1 = xla_hlo.add %arg1, %arg1 : tensor<4xf32>
+// CHECK-LABEL: func @sameBenefit
+func @sameBenefit(%arg0 : tensor<4xf32>) -> tensor<4xf32> {
+  // Because these are all the same benefit, initial formation puts them each
+  // in their own region.
+  // CHECK: flow.dispatch.region
+  // CHECK:   xla_hlo.add
+  // CHECK: flow.dispatch.region
+  // CHECK:   call @callee
```

Comment:
why aren't these together now?

Reply:
Because they are the same benefit and I removed the output fusing workaround because it was violating invariants of the IR in some cases, wasn't general to dynamic shapes, and was tricky to get right/highly restricted anyway.

I also fiddled with traversing in reverse order (like the original) but that made it very hard to not swallow too much, triggering some of the same bad behavior as the original. When I looked at the effect of this pass and the current MergeCompatibleDispatchRegions, I decided that we would be better off having a pass (this one) that was primarily about identifying the most important highly parallel anchors and forming regions around them in benefit order, letting us then identify the parallelism, do the CSE/analysis to merge, then have optimization pass(es) to take it the rest of the way (which is fairly open ended and really needs a cost model to be great).

Even with it as it is in this new state, some things are different, but when I spot checked the various dispatch regions, it was recovering plausible sequences. Given that, the fact that a lot of this stuff still needs work, and when tensor-fusion lands, the problem changes pretty dramatically, I opted for this simplification as I can see better ways forward from it.

(and if not, we still have the previous version :) )